### PR TITLE
77741 Complete endpoint and update attendance status and notes

### PR DESCRIPTION
**Pull request by bratli**
DevOps#77741

Endpoint do complete punch out. The contractor can also update attendance status and notes on the participants. This can be done in the Complete endpoint, or as a separate endpoint (as long as the IPO is in completed stage)

### CODE REVIEW

File: src/Equinor.ProCoSys.IPO.MainApi/MainApi/Person/MainApiPersonService.cs [hunk 1]
```diff
@@ -81,5 +81,22 @@ public async Task<ProCoSysPerson> GetPersonByOidsInUserGroupAsync(string plant,
...
+            string azureOid,
+            string functionalRoleCode)
+        {
+            var url = $"{_baseAddress}Person/PersonByOidInFunctionalRole" +
+                      $"?plantId={plant}" +
+                      $"&azureOid={WebUtility.UrlEncode(azureOid)}" +
+                      $"&functionalRoleCode={WebUtility.UrlEncode(functionalRoleCode)}" +
+                      $"&api-version={_apiVersion}";
+
+            var person = await _foreignApiClient.QueryAndDeserializeAsync<ProCoSysPerson>(url);
```
Comment (llykkelarsen) [inline src/Equinor.ProCoSys.IPO.MainApi/MainApi/Person/MainApiPersonService.cs:96]: return await, no need for variable

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs [hunk 2]
```diff
@@ -95,6 +97,43 @@ public class InvitationsController : ControllerBase
             return this.FromResult(result);
         }
 
+        [Authorize(Roles = Permissions.IPO_SIGN)]
+        [Authorize(Roles = Permissions.IPO_WRITE)]
```
Comment (llykkelarsen) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs:101]: We need to implement functionality for several roles as has been done in preservation. See for example TagsController ln 77 in preservation, the AuthorizeAny method. We can talk about it if unclear :)

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs [hunk 3]
```diff
@@ -95,6 +97,43 @@ public class InvitationsController : ControllerBase
...
+        [HttpPut("{id}/Complete")]
+        public async Task<ActionResult<string>> CompleteInvitation(
+            [FromHeader(Name = CurrentPlantMiddleware.PlantHeader)]
+            [Required]
+            [StringLength(PlantEntityBase.PlantLengthMax, MinimumLength = PlantEntityBase.PlantLengthMin)]
+            string plant,
+            [FromRoute] int id,
+            [FromBody] CompleteInvitationDto dto)
+        {
+            var participantsToChange = dto.Participants.Select(p =>
```
Comment (llykkelarsen) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs:111]: participantsToUpdate?

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/CompleteInvitation/CompleteInvitationCommandHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                true,
+                _note,
+                _participantRowVersion1),
+            new UpdateAttendedStatusAndNotesOnParticipantsForCommand(
+                _participantId2,
+                true,
+                _note,
+                _participantRowVersion2)
+        };
+
```
Comment (llykkelarsen) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/CompleteInvitation/CompleteInvitationCommandHandlerTests.cs:60]: whitespace

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/CompleteInvitation/CompleteInvitationCommandHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            _unitOfWorkMock = new Mock<IUnitOfWork>();
+            _unitOfWorkMock
+                .Setup(x => x.SaveChangesAsync(It.IsAny<CancellationToken>()))
+                .Callback(() => _saveChangesCount++);
+
+            _currentUserProviderMock = new Mock<ICurrentUserProvider>();
+            _currentUserProviderMock
+                .Setup(x => x.GetCurrentUserOid()).Returns(_azureOidForCurrentUser);
+
```
Comment (llykkelarsen) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/CompleteInvitation/CompleteInvitationCommandHandlerTests.cs:94]: whitespace

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/CompleteInvitation/CompleteInvitationCommandHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            _dut = new CompleteInvitationCommandHandler(
+                _plantProviderMock.Object,
+                _invitationRepositoryMock.Object,
+                _unitOfWorkMock.Object,
+                _currentUserProviderMock.Object,
+                _personApiServiceMock.Object);
+        }
+
+        [TestMethod]
+        public async Task CompleteIpoIpoCommand_ShouldCompleteInvitation()
```
Comment (llykkelarsen) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/CompleteInvitation/CompleteInvitationCommandHandlerTests.cs:162]: CompleteIpoCommand?

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/EditInvitation/EditInvitationCommandHandlerTests.cs [hunk 7]
```diff
@@ -3,6 +3,7 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Castle.DynamicProxy.Generators.Emitters.SimpleAST;
```
Comment (llykkelarsen) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/EditInvitation/EditInvitationCommandHandlerTests.cs:6]: remove

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/UpdateAttendedStatusAndNotesOnParticipants/UpdateAttendedStatusAndNotesOnParticipantsCommandValidatorTests.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+            var result = _dut.Validate(_command);
+
+            Assert.IsFalse(result.IsValid);
+            Assert.AreEqual(1, result.Errors.Count);
+            Assert.IsTrue(result.Errors[0].ErrorMessage.StartsWith("Invitation is not in completed stage, and thus cannot change attended statuses"));
+        }
+
+        [TestMethod]
+        public void Validate_ShouldFail_WhenRowVersionIsInvalid()
```
Comment (llykkelarsen) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/InvitationCommands/UpdateAttendedStatusAndNotesOnParticipants/UpdateAttendedStatusAndNotesOnParticipantsCommandValidatorTests.cs:107]: duplicate test?

**Review by llykkelarsen — CHANGES_REQUESTED** [review]

File: src/Equinor.ProCoSys.IPO.Command/InvitationCommands/UpdateAttendedStatusAndNotesOnParticipants/UpdateAttendedStatusAndNotesOnParticipantsCommandHandler.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                p.Organization == Organization.Contractor &&
+                p.AzureOid == currentUserAzureOid).ToList();
+
+            if (!participants.Any() || participants[0].FunctionalRoleCode != null)
+            {
+                var functionalRole = invitation.Participants
+                    .SingleOrDefault(p => p.SortKey == 0 &&
+                                          p.FunctionalRoleCode != null &&
+                                          p.Type == IpoParticipantType.FunctionalRole);
+
```
[the remark this answers is not in the document]
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.Command/InvitationCommands/UpdateAttendedStatusAndNotesOnParticipants/UpdateAttendedStatusAndNotesOnParticipantsCommandHandler.cs:51]: Change to .Single. We check for this in the validator, so it shouldnt ever "crash"

File: src/Equinor.ProCoSys.IPO.Command/InvitationCommands/CompleteInvitation/CompleteInvitationCommandValidator.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+                    $"Invitation row version is not valid! InvitationRowVersion={command.InvitationRowVersion}")
+                .Must(command => HaveAValidRowVersion(command.ParticipantRowVersion))
+                .WithMessage(command =>
+                    $"Participant row version is not valid! ParticipantRowVersion={command.ParticipantRowVersion}")
+                .MustAsync((command, token) => BeAContractorOnIpo(command.InvitationId, token))
+                .WithMessage(command =>
+                    "The IPO does not have a contractor assigned to complete the IPO!")
+                .MustAsync((command, token) => BeTheAssignedPersonIfPersonParticipant(command.InvitationId, token))
+                .WithMessage(command =>
+                    "Person signing is not the contractor assigned to complete this IPO, or there is not a valid functional role on the IPO!");
```
[the remark this answers is not in the document]
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.Command/InvitationCommands/CompleteInvitation/CompleteInvitationCommandValidator.cs:34]: It is hard to split this up, but the test checks for scenarios that should never happen. Should I drop this? Alternatively I can change the message to just "Person signing is not the contractor assigned to complete this IPO", and then this message will also appear if say for some reason there was an IPO with contractor: person with FunctionalRoleCode, but there is no row in db for the actual FR (type = FunctionalRole)

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs (hunk 2 above)
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs:101]: Talked to JI, the way it is now is correct when the user needs both of these permissions

File: src/Equinor.ProCoSys.IPO.Command/InvitationCommands/CompleteInvitation/CompleteInvitationCommandValidator.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (llykkelarsen) [reply src/Equinor.ProCoSys.IPO.Command/InvitationCommands/CompleteInvitation/CompleteInvitationCommandValidator.cs]: Oh I see now. Uh I guess it's not great having an error message so far from what the error actually is either. I say keep it as is or drop that part of the validation.

**Review by llykkelarsen — APPROVED** [review]

**Comment by llykkelarsen** [comment]
Forgot to share this thought yesterday: the wording 'complete invitation', should it rather be 'complete punch out round'/'complete punch out'/something like that? We're not really completing the invitation, are we? Irrelevant for functionality obviously but just thought I'd ask

**Pull request merged**